File: internal/handlers/bluetooth.go
package handlers

import (
	"net/http"

	"github.com/labstack/echo/v4"
	"github.com/nerzhul/home-bt-broker/internal/bluetooth"
)

// BluetoothHandler handles Bluetooth-related endpoints
type BluetoothHandler struct {
	btManager *bluetooth.BluetoothManager
}

// NewBluetoothHandler creates a new Bluetooth handler
func NewBluetoothHandler() (*BluetoothHandler, error) {
	btManager, err := bluetooth.NewBluetoothManager()
	if err != nil {
		return nil, err
	}

	return &BluetoothHandler{btManager: btManager}, nil
}

// Close closes the Bluetooth manager connection
func (bh *BluetoothHandler) Close() {
	if bh.btManager != nil {
		bh.btManager.Close()
	}
}

// GetAdapters returns all Bluetooth adapters
func (bh *BluetoothHandler) GetAdapters(c echo.Context) error {
	adapters, err := bh.btManager.GetAdapters()
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{
			"error": "failed to get adapters: " + err.Error(),
		})
	}

	return c.JSON(http.StatusOK, map[string]interface{}{
		"adapters": adapters,
	})
}

// GetDevices returns all devices for a specific adapter by MAC address
func (bh *BluetoothHandler) GetDevices(c echo.Context) error {
	adapterMAC := c.Param("adapter")
	if adapterMAC == "" {
		return c.JSON(http.StatusBadRequest, map[string]string{
			"error": "adapter MAC address parameter is required",
		})
	}

	// Resolve MAC address to adapter path
	adapterPath, err := bh.btManager.GetAdapterPathByMAC(adapterMAC)
	if err != nil {
		return c.JSON(http.StatusNotFound, map[string]string{
			"error": "adapter not found: " + err.Error(),
		})
	}

	devices, err := bh.btManager.GetDevices(adapterPath)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{
			"error": "failed to get devices: " + err.Error(),
		})
	}

	return c.JSON(http.StatusOK, map[string]interface{}{
		"devices": devices,
	})
}

// GetTrustedDevices returns trusted devices for a specific adapter by MAC address
func (bh *BluetoothHandler) GetTrustedDevices(c echo.Context) error {
	adapterMAC := c.Param("adapter")
	if adapterMAC == "" {
		return c.JSON(http.StatusBadRequest, map[string]string{
			"error": "adapter MAC address parameter is required",
		})
	}

	// Resolve MAC address to adapter path
	adapterPath, err := bh.btManager.GetAdapterPathByMAC(adapterMAC)
	if err != nil {
		return c.JSON(http.StatusNotFound, map[string]string{
			"error": "adapter not found: " + err.Error(),
		})
	}

	devices, err := bh.btManager.GetTrustedDevices(adapterPath)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{
			"error": "failed to get trusted devices: " + err.Error(),
		})
	}

	return c.JSON(http.StatusOK, map[string]interface{}{
		"trusted_devices": devices,
	})
}

// GetConnectedDevices returns connected devices for a specific adapter by MAC address
func (bh *BluetoothHandler) GetConnectedDevices(c echo.Context) error {
	adapterMAC := c.Param("adapter")
	if adapterMAC == "" {
		return c.JSON(http.StatusBadRequest, map[string]string{
			"error": "adapter MAC address parameter is required",
		})
	}

	// Resolve MAC address to adapter path
	adapterPath, err := bh.btManager.GetAdapterPathByMAC(adapterMAC)
	if err != nil {
		return c.JSON(http.StatusNotFound, map[string]string{
			"error": "adapter not found: " + err.Error(),
		})
	}

	devices, err := bh.btManager.GetConnectedDevices(adapterPath)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{
			"error": "failed to get connected devices: " + err.Error(),
		})
	}

	return c.JSON(http.StatusOK, map[string]interface{}{
		"connected_devices": devices,
	})
}

// ConnectDevice connects to a device by MAC address using adapter MAC
func (bh *BluetoothHandler) ConnectDevice(c echo.Context) error {
	adapterMAC := c.Param("adapter")
	macAddress := c.Param("mac")
	
	if adapterMAC == "" {
		return c.JSON(http.StatusBadRequest, map[string]string{
			"error": "adapter MAC address parameter is required",
		})
	}
	
	if macAddress == "" {
		return c.JSON(http.StatusBadRequest, map[string]string{
			"error": "device MAC address parameter is required",
		})
	}

	// Resolve MAC address to adapter path
	adapterPath, err := bh.btManager.GetAdapterPathByMAC(adapterMAC)
	if err != nil {
		return c.JSON(http.StatusNotFound, map[string]string{
			"error": "adapter not found: " + err.Error(),
		})
	}

	err = bh.btManager.ConnectDevice(adapterPath, macAddress)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{
			"error": "failed to connect device: " + err.Error(),
		})
	}

	return c.JSON(http.StatusOK, map[string]string{
		"message": "device connection initiated successfully",
	})
}

// TrustDevice trusts a device by MAC address using adapter MAC
func (bh *BluetoothHandler) TrustDevice(c echo.Context) error {
	adapterMAC := c.Param("adapter")
	macAddress := c.Param("mac")
	
	if adapterMAC == "" {
		return c.JSON(http.StatusBadRequest, map[string]string{
			"error": "adapter MAC address parameter is required",
		})
	}
	
	if macAddress == "" {
		return c.JSON(http.StatusBadRequest, map[string]string{
			"error": "device MAC address parameter is required",
		})
	}

	// Resolve MAC address to adapter path
	adapterPath, err := bh.btManager.GetAdapterPathByMAC(adapterMAC)
	if err != nil {
		return c.JSON(http.StatusNotFound, map[string]string{
			"error": "adapter not found: " + err.Error(),
		})
	}

	err = bh.btManager.TrustDevice(adapterPath, macAddress)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{
			"error": "failed to trust device: " + err.Error(),
		})
	}

	return c.JSON(http.StatusOK, map[string]string{
		"message": "device trusted successfully",
	})
}

// RemoveDevice removes a device by MAC address using adapter MAC
func (bh *BluetoothHandler) RemoveDevice(c echo.Context) error {
	adapterMAC := c.Param("adapter")
	macAddress := c.Param("mac")
	
	if adapterMAC == "" {
		return c.JSON(http.StatusBadRequest, map[string]string{
			"error": "adapter MAC address parameter is required",
		})
	}
	
	if macAddress == "" {
		return c.JSON(http.StatusBadRequest, map[string]string{
			"error": "device MAC address parameter is required",
		})
	}

	// Resolve MAC address to adapter path
	adapterPath, err := bh.btManager.GetAdapterPathByMAC(adapterMAC)
	if err != nil {
		return c.JSON(http.StatusNotFound, map[string]string{
			"error": "adapter not found: " + err.Error(),
		})
	}

	err = bh.btManager.RemoveDevice(adapterPath, macAddress)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{
			"error": "failed to remove device: " + err.Error(),
		})
	}

	return c.JSON(http.StatusOK, map[string]string{
		"message": "device removed successfully",
	})
}

// PairDevice pairs with a device by MAC address using adapter MAC
func (bh *BluetoothHandler) PairDevice(c echo.Context) error {
	adapterMAC := c.Param("adapter")
	macAddress := c.Param("mac")
	
	if adapterMAC == "" {
		return c.JSON(http.StatusBadRequest, map[string]string{
			"error": "adapter MAC address parameter is required",
		})
	}
	
	if macAddress == "" {
		return c.JSON(http.StatusBadRequest, map[string]string{
			"error": "device MAC address parameter is required",
		})
	}

	// Resolve MAC address to adapter path
	adapterPath, err := bh.btManager.GetAdapterPathByMAC(adapterMAC)
	if err != nil {
		return c.JSON(http.StatusNotFound, map[string]string{
			"error": "adapter not found: " + err.Error(),
		})
	}

	err = bh.btManager.PairDevice(adapterPath, macAddress)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{
			"error": "failed to pair device: " + err.Error(),
		})
	}

	return c.JSON(http.StatusOK, map[string]string{
		"message": "device pairing initiated successfully",
	})
}

File: src/handlers/bluetooth.rs
use axum::{
    extract::{Path, State},
    http::StatusCode,
    response::{IntoResponse, Response},
    Json,
};
use serde::Deserialize;

use super::err_json;
use crate::AppState;

pub async fn get_adapters(State(state): State<AppState>) -> Response {
    match state.bluetooth.get_adapters().await {
        Ok(adapters) => Json(serde_json::json!({ "adapters": adapters })).into_response(),
        Err(e) => err_json(
            StatusCode::INTERNAL_SERVER_ERROR,
            &format!("failed to get adapters: {}", e),
        ),
    }
}

pub async fn get_devices(
    State(state): State<AppState>,
    Path(adapter): Path<String>,
) -> Response {
    let path = match state.bluetooth.get_adapter_path_by_mac(&adapter).await {
        Ok(p) => p,
        Err(e) => return err_json(StatusCode::NOT_FOUND, &format!("adapter not found: {}", e)),
    };
    match state.bluetooth.get_devices(&path).await {
        Ok(devices) => Json(serde_json::json!({ "devices": devices })).into_response(),
        Err(e) => err_json(
            StatusCode::INTERNAL_SERVER_ERROR,
            &format!("failed to get devices: {}", e),
        ),
    }
}

pub async fn get_trusted_devices(
    State(state): State<AppState>,
    Path(adapter): Path<String>,
) -> Response {
    let path = match state.bluetooth.get_adapter_path_by_mac(&adapter).await {
        Ok(p) => p,
        Err(e) => return err_json(StatusCode::NOT_FOUND, &format!("adapter not found: {}", e)),
    };
    match state.bluetooth.get_trusted_devices(&path).await {
        Ok(devices) => Json(serde_json::json!({ "trusted_devices": devices })).into_response(),
        Err(e) => err_json(
            StatusCode::INTERNAL_SERVER_ERROR,
            &format!("failed to get trusted devices: {}", e),
        ),
    }
}

pub async fn get_connected_devices(
    State(state): State<AppState>,
    Path(adapter): Path<String>,
) -> Response {
    let path = match state.bluetooth.get_adapter_path_by_mac(&adapter).await {
        Ok(p) => p,
        Err(e) => return err_json(StatusCode::NOT_FOUND, &format!("adapter not found: {}", e)),
    };
    match state.bluetooth.get_connected_devices(&path).await {
        Ok(devices) => {
            Json(serde_json::json!({ "connected_devices": devices })).into_response()
        }
        Err(e) => err_json(
            StatusCode::INTERNAL_SERVER_ERROR,
            &format!("failed to get connected devices: {}", e),
        ),
    }
}

#[derive(Deserialize)]
pub struct EnableRequest {
    pub enable: bool,
}

pub async fn set_discoverable(
    State(state): State<AppState>,
    Path(adapter): Path<String>,
    Json(req): Json<EnableRequest>,
) -> Response {
    let path = match state.bluetooth.get_adapter_path_by_mac(&adapter).await {
        Ok(p) => p,
        Err(e) => return err_json(StatusCode::NOT_FOUND, &format!("adapter not found: {}", e)),
    };
    match state.bluetooth.set_discoverable(&path, req.enable).await {
        Ok(_) => Json(serde_json::json!({ "message": "discoverable updated" })).into_response(),
        Err(e) => err_json(
            StatusCode::INTERNAL_SERVER_ERROR,
            &format!("failed to set discoverable: {}", e),
        ),
    }
}

pub async fn set_discovering(
    State(state): State<AppState>,
    Path(adapter): Path<String>,
    Json(req): Json<EnableRequest>,
) -> Response {
    let path = match state.bluetooth.get_adapter_path_by_mac(&adapter).await {
        Ok(p) => p,
        Err(e) => return err_json(StatusCode::NOT_FOUND, &format!("adapter not found: {}", e)),
    };
    match state.bluetooth.set_discovering(&path, req.enable).await {
        Ok(_) => Json(serde_json::json!({ "message": "discovering updated" })).into_response(),
        Err(e) => err_json(
            StatusCode::INTERNAL_SERVER_ERROR,
            &format!("failed to set discovering: {}", e),
        ),
    }
}

pub async fn connect_device(
    State(state): State<AppState>,
    Path((adapter, mac)): Path<(String, String)>,
) -> Response {
    let path = match state.bluetooth.get_adapter_path_by_mac(&adapter).await {
        Ok(p) => p,
        Err(e) => return err_json(StatusCode::NOT_FOUND, &format!("adapter not found: {}", e)),
    };
    match state.bluetooth.connect_device(&path, &mac).await {
        Ok(_) => Json(serde_json::json!({
            "message": "device connection initiated successfully"
        }))
        .into_response(),
        Err(e) => err_json(
            StatusCode::INTERNAL_SERVER_ERROR,
            &format!("failed to connect device: {}", e),
        ),
    }
}

pub async fn trust_device(
    State(state): State<AppState>,
    Path((adapter, mac)): Path<(String, String)>,
) -> Response {
    let path = match state.bluetooth.get_adapter_path_by_mac(&adapter).await {
        Ok(p) => p,
        Err(e) => return err_json(StatusCode::NOT_FOUND, &format!("adapter not found: {}", e)),
    };
    match state.bluetooth.trust_device(&path, &mac).await {
        Ok(_) => {
            Json(serde_json::json!({ "message": "device trusted successfully" })).into_response()
        }
        Err(e) => err_json(
            StatusCode::INTERNAL_SERVER_ERROR,
            &format!("failed to trust device: {}", e),
        ),
    }
}

pub async fn pair_device(
    State(state): State<AppState>,
    Path((adapter, mac)): Path<(String, String)>,
) -> Response {
    let path = match state.bluetooth.get_adapter_path_by_mac(&adapter).await {
        Ok(p) => p,
        Err(e) => return err_json(StatusCode::NOT_FOUND, &format!("adapter not found: {}", e)),
    };
    match state.bluetooth.pair_device(&path, &mac).await {
        Ok(_) => Json(serde_json::json!({
            "message": "device pairing initiated successfully"
        }))
        .into_response(),
        Err(e) => err_json(
            StatusCode::INTERNAL_SERVER_ERROR,
            &format!("failed to pair device: {}", e),
        ),
    }
}

pub async fn remove_device(
    State(state): State<AppState>,
    Path((adapter, mac)): Path<(String, String)>,
) -> Response {
    let path = match state.bluetooth.get_adapter_path_by_mac(&adapter).await {
        Ok(p) => p,
        Err(e) => return err_json(StatusCode::NOT_FOUND, &format!("adapter not found: {}", e)),
    };
    match state.bluetooth.remove_device(&path, &mac).await {
        Ok(_) => {
            Json(serde_json::json!({ "message": "device removed successfully" })).into_response()
        }
        Err(e) => err_json(
            StatusCode::INTERNAL_SERVER_ERROR,
            &format!("failed to remove device: {}", e),
        ),
    }
}
